actions/test/run_local.ts

```diff
@@ -72,58 +87,124 @@ async function processBlock(
   );
   let hasErrors = false;
   for (const transaction of blockWithTransactions.transactions) {
-    const receipt = await provider.getTransactionReceipt(transaction.hash);
-    if (receipt) {
-      const {
-        hash,
-        from,
-        value,
-        nonce,
-        gasLimit,
-        maxPriorityFeePerGas,
-        maxFeePerGas,
-      } = transaction;
-
-      const testTransactionEvent: TestTransactionEvent = {
-        blockHash: block.hash,
-        blockNumber: block.number,
-        from,
-        hash,
-        network: chainId.toString(),
-        logs: receipt.logs,
-        input: "",
-        value: value.toString(),
-        nonce: nonce.toString(),
-        gas: gasLimit.toString(),
-        gasUsed: receipt.gasUsed.toString(),
-        cumulativeGasUsed: receipt.cumulativeGasUsed.toString(),
-        gasPrice: receipt.effectiveGasPrice.toString(),
-        gasTipCap: maxPriorityFeePerGas ? maxPriorityFeePerGas.toString() : "",
-        gasFeeCap: maxFeePerGas ? maxFeePerGas.toString() : "",
-        transactionHash: transaction.hash,
-      };
-
-      // run action
-      console.log(`[run_local] Run "addContract" action for TX ${hash}`);
-      const result = await testRuntime
-        .execute(addContract, testTransactionEvent)
-        .then(() => true)
-        .catch((e) => {
-          hasErrors = true;
-          console.error(
-            `[run_local] Error running "addContract" action for TX:`,
-            e
-          );
-          return false;
-        });
-      console.log(
-        `[run_local] Result of "addContract" action for TX ${hash}: ${_formatResult(
-          result
-        )}`
-      );
-    }
+    hasErrors ||= !(await !_processTx(
+      provider,
+      block,
+      chainId,
+      testRuntime,
+      transaction
+    ));
+  }
+
+  hasErrors ||= !(await _pollAndPost({ block, chainId, testRuntime }));
+
+  if (hasErrors) {
+    throw new Error("[run_local] Errors found in processing block");
+  }
+}
+
+async function processTx(
+  provider: ethers.providers.Provider,
+  tx: string,
+  chainId: number,
+  testRuntime: TestRuntime
+) {
+  let hasErrors = false;
+  const transaction = await provider.getTransaction(tx);
+  if (!transaction.blockNumber) {
+    throw new Error(`The transaction ${tx} is not mined yet (no blockNumber)`);
+  }
+  const block = await provider.getBlock(transaction.blockNumber);
+  if (!transaction) {
```

Comment:
If the transaction can be falsy, line 114 can fail. Better add the `?.` operator.
Or move the `transaction` check up

```suggestion
  const transaction = await provider.getTransaction(tx);
  if (!transaction?.blockNumber) {
    throw new Error(`The transaction ${tx} is not mined yet (no blockNumber)`);
  }
  const block = await provider.getBlock(transaction?.blockNumber);
  if (!transaction) {
```

Reply:
But why would it be falsy? it will throw if there's no TX, if there is, it will give me the receipt. Right?

Reply:
successful promise will give the Tx Response...